ddht/v5_1/alexandria/network.py

```diff
@@ -230,6 +217,11 @@ def recursive_find_nodes(
     ) -> AsyncContextManager[trio.abc.ReceiveChannel[ENRAPI]]:
         return common_recursive_find_nodes(self, target)
 
+    def explore(
+        self, target: NodeID, concurrency: int = 3,
+    ) -> AsyncContextManager[trio.abc.ReceiveChannel[ENRAPI]]:
+        return common_explore_network(self, target, concurrency=concurrency)
```

Comment:
I'm not exactly sure where your confusion lies but I think the short answer is yes.

The long answer:

```python
@asynccontextmanager
async def something() -> AsyncIterator[None]:
    do_something_before()
    yield
    do_something_after()
```

The `AsyncIterator` is to satisfy `mypy`.  The decorator transforms it roughly like this:

```python
TContextValue = TypeVar("TContextValue")

def asynccontextmanager(func: Callable[..., AsyncIterator[TContextValue]]) -> AsyncContextManager[TContextValue]:
    ...
```

So the decorator uses the `TypeVar` approach to transform the return type of the function.

As for why you're not seeing any `async` in the function definitions.

```python
@asynccontextmanager
async def my_context_fn() -> AsyncIterator[None]:
    ...

def proxy_to_my_context_fn() -> AsyncContextManager[None]:
    return my_context_fn()  # this is confusing, why don't I have to call this with `await`?????
```

The reason is that the return value of `my_context_fn()` is an instance of an object that implements the `AsyncContextManager` API:

```python
class AsyncContextManager:
    async def __aenter__(...):
    async def __aexit__(...):
```

So we don't need the `await` on the function call because python does it for us when we use it like `async with proxy_to_my_context_fn():`

So... there's no `async` needed for the definition of `NetworkAPI.explore` and no need to `await` the call to `common_explore_network`.

Reply:
This is great, cleared everything up, thanks!